Support ampersand-prefixed number literals

fourls opened the pull request:
Fixes #173 by accepting and discarding any ampersands before a number literal.

Some caveats:

- This occurs during parsing, not lexing (which might be more correct). This is because our ampersand handling for identifiers is in the parser. 
- This grammar will technically support whitespace between the ampersands and the number literals. I don't think this should be an issue as there's no ambiguity with any other language constructs.

Cirras reviewed (changes requested):
> This occurs during parsing, not lexing (which might be more correct). This is because our ampersand handling for identifiers is in the parser.

I'd strongly prefer handling this in the lexer.
We can make that same change to identifiers, unless you think there's a good reason not to.

This would also involve deprecating the `AMPERSAND` token in the grammar (just change the value from `'&'` to `'@AMPERSAND@'`.

I'll come along later and start generating deprecations for `DelphiTokenType` enum values based on some sort of token naming conventions, but we can leave that for now.

fourls commented:
I have:
- Moved the handling of ampersands for identifiers, int literals, and real literals down into the lexer
- Overridden `IdentifierNodeImpl.getImage` to strip up to one leading ampersand
- Modified `IntLiteralNodeImpl.getValue` to handle ampersands in the image
- Modified `RealLiteralNodeImpl.getValue` to handle ampersands in the image
- Modified `DigitGroupingCheck` to handle ampersands in literal images

This seems to work. I've added some extra testing, and the existing symbol table test at `DelphiSymbolTableExecutorTest.testAmpersands` passes.

Cirras commented on delphi-checks/src/test/java/au/com/integradev/delphi/checks/DigitGroupingCheckTest.java:
```diff
@@ -36,6 +36,10 @@ class DigitGroupingCheckTest {
         "$ABC_DEF",
         "$ABC_DE_F",
         "%1_001_01",
+        "&1_001_01",
```
nit: these look like a binary literal at first, can you change the non-leading 1s to 0?

Cirras commented on delphi-checks/src/test/java/au/com/integradev/delphi/checks/DigitGroupingCheckTest.java:
```diff
@@ -62,6 +66,10 @@ void testNonStandardGroupingShouldAddIssue(String actual) {
         "%1_01_01",
         "%10_101",
         "%10_1010",
+        "&100_101",
```
nit: these look like a binary literal at first, can you change the non-leading 1s to 0?

Cirras commented on delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g:
```diff
@@ -81,6 +81,15 @@ package au.com.integradev.delphi.antlr;
     return "line " + e.line + ":" + e.charPositionInLine;
   }
 
+  private boolean nextNonAmpersandIsDigit() {
```
Unused function.

Cirras commented on delphi-frontend/src/test/java/au/com/integradev/delphi/antlr/GrammarTest.java:
```diff
@@ -262,6 +262,11 @@ void testUnusualIntegerLiterals() {
   }
 
+  @Test
+  void testAmpersandNumberLiterals() {
```
I feel like we normally say "numeric literals" instead of "number literals".

Cirras reviewed (approved):
Well done! 👍
